sycl/include/sycl/group_algorithm.hpp

```diff
@@ -59,29 +59,22 @@ template <> inline id<3> linear_id_to_id(range<3> r, size_t linear_id) {
 }
 
 // ---- get_local_linear_range
-template <typename Group> size_t get_local_linear_range(Group g);
-template <> inline size_t get_local_linear_range<group<1>>(group<1> g) {
-  return g.get_local_range(0);
-}
-template <> inline size_t get_local_linear_range<group<2>>(group<2> g) {
-  return g.get_local_range(0) * g.get_local_range(1);
-}
-template <> inline size_t get_local_linear_range<group<3>>(group<3> g) {
-  return g.get_local_range(0) * g.get_local_range(1) * g.get_local_range(2);
-}
-template <>
-inline size_t get_local_linear_range<sycl::sub_group>(sycl::sub_group g) {
-  return g.get_local_range()[0];
+template <typename Group> inline auto get_local_linear_range(Group g) {
+  auto local_range = g.get_local_range();
+  auto result = local_range[0];
+  for (size_t i = 1; i < Group::dimensions; ++i)
+    result *= local_range[i];
+  return result;
 }
 
 // ---- get_local_linear_id
-template <typename Group>
-inline typename Group::linear_id_type get_local_linear_id(Group g);
+template <typename Group> inline auto get_local_linear_id(Group g) {
+  return g.get_local_id();
+}
```

Comment:
This is very confusing to read. Signature is "linear" but we return just "local" and there is no SFINAE/etc...

Reply:
You are right. Since it is now auto it can be a little confusing as you don't get the implicit conversion to scalar directly. I have addressed it.